src/dpo/src/Optdp.cpp

```diff
@@ -0,0 +1,855 @@
+      Node* nd = it_n->second;
+      // XXX: Need to figure out orientation.
+      int lx = (int)(nd->getX() - 0.5 * nd->getWidth());
+      int yb = (int)(nd->getY() - 0.5 * nd->getHeight());
+      inst->setLocation(lx, yb);
+
+      double xmin = nd->getX() - 0.5 * nd->getWidth();
+      double xmax = nd->getX() + 0.5 * nd->getWidth();
+      double ymin = nd->getY() - 0.5 * nd->getHeight();
+      double ymax = nd->getY() + 0.5 * nd->getHeight();
+    }
+  }
+}
+////////////////////////////////////////////////////////////////
+void Optdp::initEdgeTypes() {
+  // Do nothing.  Use padding instead.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::initCellSpacingTable() {
+  // Do nothing.  Use padding instead.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::initPadding() {
+  // Grab information from OpenDP.
+  dpl::Opendp* opendp = openroad_->getOpendp();
+
+  // Need to turn off spacing tables and turn on padding.
+  arch_->setUseSpacingTable(false);
+  arch_->setUsePadding(true);
+  arch_->init_edge_type();
+
+
+  // Create and edge type for each amount of padding.  This
+  // can be done by querying OpenDP.
+  dbSet<dbRow> rows = db_->getChip()->getBlock()->getRows();
+  if (rows.empty()) {
+    return;
+  }
+  int siteWidth = (*rows.begin())->getSite()->getWidth();
+  std::unordered_map<odb::dbInst*, Node*>::iterator it_n;
+
+  dbSet<dbInst> insts = db_->getChip()->getBlock()->getInsts();
+  for (dbInst* inst : insts) {
+    it_n = instMap_.find(inst);
+    if (instMap_.end() != it_n) {
+      Node* ndi = it_n->second;
+      int leftPadding = opendp->padLeft(inst);
+      int rightPadding = opendp->padRight(inst);
+      arch_->addCellPadding(ndi, leftPadding * siteWidth,
+                            rightPadding * siteWidth);
+    }
+  }
+}
+////////////////////////////////////////////////////////////////
+void Optdp::createLayerMap() {
+  // Relates to pin blockages, etc. Maybe not needed right now.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::createNdrMap() {
+  // Relates to pin blockages, etc. Maybe not needed right now.
+  ;
+}
+////////////////////////////////////////////////////////////////
+void Optdp::setupMasterPowers() {
+  // Need to try and figure out which voltages are on the
+  // top and bottom of the masters/insts in order to set
+  // stuff up for proper row alignment of multi-height
+  // cells.  What I do it look at the individual ports
+  // (MTerms) and see which ones correspond to POWER and
+  // GROUND and then figure out which one is on top and
+  // which one is on bottom.  I also record the layers
+  // and use that information later when setting up the
+  // row powers.
+  dbBlock* block = db_->getChip()->getBlock();
+  std::vector<dbMaster*> masters;
+  block->getMasters(masters);
+
+  for (size_t i = 0; i < masters.size(); i++) {
+    dbMaster* master = masters[i];
```

Comment:
More modern style (when you don't need the index):
for (dbMaster* master : masters) {
...